MediaQueryMatcher should be a ContextDestructionObserver

AlexsanderDamaceno opened the pull request:
#### 49a60d7b038f52940ba3b197d6a0ab9806dc4dde
<pre>
MediaQueryMatcher should be a ContextDestructionObserver
<a href="https://bugs.webkit.org/show_bug.cgi?id=99240">https://bugs.webkit.org/show_bug.cgi?id=99240</a>

Reviewed by NOBODY (OOPS!).

MediaQueryMatcher should be a ContextDestructionObserver

* Source/WebCore/css/MediaQueryMatcher.cpp:
(WebCore::MediaQueryMatcher::MediaQueryMatcher):
(WebCore::MediaQueryMatcher::contextDestroyed):
(WebCore::MediaQueryMatcher::document const):
(WebCore::MediaQueryMatcher::mediaType const):
(WebCore::MediaQueryMatcher::evaluate):
(WebCore::MediaQueryMatcher::matchMedia):
(WebCore::MediaQueryMatcher::evaluateAll):
(WebCore::MediaQueryMatcher::documentDestroyed): Deleted.
* Source/WebCore/css/MediaQueryMatcher.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::willBeRemovedFromFrame):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/49a60d7b038f52940ba3b197d6a0ab9806dc4dde

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/174411 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/48344 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/42125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/183988 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/132279 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/176276 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/49636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/48026 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/135677 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/95730 "4 flakes 4 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/177369 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/39114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/156468 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/116155 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/37755 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/36119 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/32469 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/148178 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/35960 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/186414 "Built successfully") | 
| | [❌ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/39631 "Found 2 new failures in css/MediaQueryMatcher.cpp") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/40316 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/144589 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/48011 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/144447 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/48690 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/32580 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/114244 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/39348 "Passed tests") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/120641 "Found 2 new failures in css/MediaQueryMatcher.cpp") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/47197 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/10924 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/46599 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/46830 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/46657 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AlexsanderDamaceno committed:
MediaQueryMatcher should be a ContextDestructionObserver
https://bugs.webkit.org/show_bug.cgi?id=99240

Reviewed by NOBODY (OOPS!).

MediaQueryMatcher should be a ContextDestructionObserver

* Source/WebCore/css/MediaQueryMatcher.cpp:
(WebCore::MediaQueryMatcher::MediaQueryMatcher):
(WebCore::MediaQueryMatcher::mediaType const):
(WebCore::MediaQueryMatcher::documentElementUserAgentStyle const):
(WebCore::MediaQueryMatcher::evaluate):
(WebCore::MediaQueryMatcher::matchMedia):
(WebCore::MediaQueryMatcher::evaluateAll):
* Source/WebCore/css/MediaQueryMatcher.h:
(WebCore::MediaQueryMatcher::contextDestroyed):
(WebCore::MediaQueryMatcher::documentDestroyed): Deleted.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/905c224e0f48c59c591606caa2f2255d54d7aec2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2e1dd32795a207d8537fb342fee0c58d3c1f62ae. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

cdumez commented on Source/WebCore/css/MediaQueryMatcher.cpp:
```diff
@@ -39,15 +39,14 @@
 
 MediaQueryMatcher::~MediaQueryMatcher() = default;
 
 void MediaQueryMatcher::documentDestroyed()
```
This should override `contextDestroyed()` from ContextDestructionObserver instead. then we would no longer need this code in Document.cpp:
```
    if (RefPtr mediaQueryMatcher = m_mediaQueryMatcher)
        mediaQueryMatcher->documentDestroyed();
```

cdumez commented on Source/WebCore/css/MediaQueryMatcher.h:
```diff
@@ -41,7 +42,7 @@ using MediaQueryList = Vector<MediaQuery>;
 // query has changed. MediaQueryLists are invoked in the order in which they were added.
 
-class MediaQueryMatcher final : public RefCounted<MediaQueryMatcher> {
+class MediaQueryMatcher final : public ContextDestructionObserver, public RefCounted<MediaQueryMatcher> {
```
Note that you forgot to drop the m_document data member too now that you can get it from the `ContextDestructionObserver` base class.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b73ebe5e5f98d2402e4a76d62278c29ffefd6c38. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5e23b3b778ea6584d5550cd8625c989c5a21c84e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/94a9fa8bea45e8bba20f13fee85786b770a42930. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

cdumez commented:
The commit log still says "OOPS"

cdumez commented on Source/WebCore/css/MediaQueryMatcher.cpp:
```diff
@@ -39,48 +39,46 @@
-    auto mediaQueryLists = std::exchange(m_mediaQueryLists, { });
-    for (auto& mediaQueryList : mediaQueryLists) {
-        if (mediaQueryList)
-            mediaQueryList->detachFromMatcher();
```
Why did you drop this?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/10920b093bc199bbfcc7e6f630adfa22573deef9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b50b90237bf892b97b97d4dbd03430578548e766. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8517e891f19c94a72351bee6ae6b3e8d7610c3d0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37162

AlexsanderDamaceno committed:
MediaQueryMatcher should be a ContextDestructionObserver
https://bugs.webkit.org/show_bug.cgi?id=99240

Reviewed by NOBODY (OOPS!).

MediaQueryMatcher should be a ContextDestructionObserver

* Source/WebCore/css/MediaQueryMatcher.cpp:
(WebCore::MediaQueryMatcher::MediaQueryMatcher):
(WebCore::MediaQueryMatcher::contextDestroyed):
(WebCore::MediaQueryMatcher::document const):
(WebCore::MediaQueryMatcher::mediaType const):
(WebCore::MediaQueryMatcher::evaluate):
(WebCore::MediaQueryMatcher::matchMedia):
(WebCore::MediaQueryMatcher::evaluateAll):
(WebCore::MediaQueryMatcher::documentDestroyed): Deleted.
* Source/WebCore/css/MediaQueryMatcher.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::willBeRemovedFromFrame):
